sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -652,6 +653,299 @@ case class MapObjects private(
+        case ObjectType(cls) if cls.isArray && cls.getComponentType.isPrimitive =>
+          s"$loopIsNull = false"
+        case _ => s"$loopIsNull = $loopValue == null;"
+      }
+    val keyLoopNullCheck =
+      loopNullCheck(genKeyInputData, keyInputDataType, keyLoopIsNull, keyLoopValue)
+    val valueLoopNullCheck =
+      loopNullCheck(genValueInputData, valueInputDataType, valueLoopIsNull, valueLoopValue)
+
+    val constructBuilder = collClass match {
+      // Scala Map
+      case cls if classOf[scala.collection.Map[_, _]].isAssignableFrom(cls) =>
+        val builderClass = classOf[Builder[_, _]].getName
+        s"""
+          $builderClass $builderValue = ${collClass.getName}$$.MODULE$$.newBuilder();
+          $builderValue.sizeHint($dataLength);
+        """
+      // Java Map, AbstractMap => HashMap
+      case cls if classOf[java.util.Map[_, _]] == cls ||
+        classOf[java.util.AbstractMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.HashMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass($dataLength);"
+      // Java SortedMap, NavigableMap => TreeMap
+      case cls if classOf[java.util.SortedMap[_, _]] == cls ||
+        classOf[java.util.NavigableMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.TreeMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass();"
+      // Java ConcurrentMap => ConcurrentHashMap
+      case cls if classOf[java.util.concurrent.ConcurrentMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.concurrent.ConcurrentHashMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass();"
+      // Java ConcurrentNavigableMap => ConcurrentSkipListMap
+      case cls if classOf[java.util.concurrent.ConcurrentNavigableMap[_, _]] == cls =>
+        val builderClass = classOf[java.util.concurrent.ConcurrentSkipListMap[_, _]].getName
+        s"$builderClass $builderValue = new $builderClass();"
+      // Java concrete Map implementation
+      case cls =>
+        val builderClass = classOf[java.util.Map[_, _]].getName
+        // Check for constructor with capacity specification
+        if (Try(cls.getConstructor(Integer.TYPE)).isSuccess) {
+          s"$builderClass $builderValue = new ${cls.getName}($dataLength);"
+        } else {
+          s"$builderClass $builderValue = new ${cls.getName}();"
+        }
+    }
+
+    val (appendToBuilder, getBuilderResult) =
+      if (classOf[scala.collection.Map[_, _]].isAssignableFrom(collClass)) {
+        val tupleClass = classOf[(_, _)].getName
+        s"""
+          $tupleClass $tupleLoopValue;
+
+          if (${genValueFunction.isNull}) {
+            $tupleLoopValue = new $tupleClass($genKeyFunctionValue, null);
+          } else {
+            $tupleLoopValue = new $tupleClass($genKeyFunctionValue, $genValueFunctionValue);
+          }
+
+          $builderValue.$$plus$$eq($tupleLoopValue);
```

Comment:
Unfortunately, `MapBuilder` only [accepts Tuples](https://www.scala-lang.org/api/current/scala/collection/mutable/MapBuilder.html#+=(x:(A,B)):MapBuilder.this.type)